File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode *fast = head,*slow = head;
        while(fast!=NULL && fast->next !=NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        if(fast!=NULL && fast->next ==NULL)
            slow = slow->next;
        ListNode *prev = NULL,*temp = NULL;
        while(slow!=NULL && slow->next !=NULL)
        {
            temp = slow->next;
            slow->next = prev;
            prev = slow;
            slow = temp;
        }
        if(slow!=NULL)
            slow->next = prev;
        fast = head;
        while(slow && fast)
        {
            if(slow->val != fast->val)
                return false;
            else{
                slow = slow->next;
                fast = fast->next;
            }
        }
        return true;
    }
    
    
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head .next == null)
            return true;
        ListNode f = head, s = head;
        while(f!=null && f.next!=null)
        {
            f = f.next.next;
            s = s.next;
        }
        ListNode curr = s, prev  = null;
        while(curr!=null)
        {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        ListNode l = head, r = prev;
        while(r!=null)
        {
            if(l.val != r.val)
                return false;
            r = r.next;
            l = l.next;
        }
        return true;
    }
}